### DISCUSSION TITLE
Mono repo support for gradle techstack

### DISCUSSION THREAD

**Discussion by skagithala-clgx** [paketo-buildpacks/feedback#247] [Java Team]
Hi Team,

We have use case where Mono repo is configured and it is using gradle with packeto build pack and it is using common folder for dependencies resolutions.
**Issue:**
It is not able to find common folder and unable to build the docker image 
below are the configurations 
 **settings.gradle** 
rootProject.name = 'xyz'
include 'project1'
include 'project2'
where project2 is common project. 
In build.gradle of project1 below line is mentioned for dependency 
implementation project (path: ':project2', configuration: 'default')

**Project structure:**
project1
project2
gradlew
gradlew.bat
settings.gradle

Could you please guide on mono repo configuration, provide if we have any documentation for the same.

**Comment by dmikusa** [comment]
When you build a monorepo with pack you have two choices:

1. Build everything locally, i.e. run `./gradle build` (or whatever) to generate your deployment artifact (JAR or WAR file). Then run `pack build -p path/to/jar-or-war-file`. Buildpacks will run off the JAR or WAR your provide and turn that into an image.

2. Build from source with buildpacks. In this case, run `pack build -p prj_root` where `prj_root` is the root of all of your projects. You need to ship the entire mono repo into the container. Buildpacks will then run and execute gradle for you. It is going to run `./gradlew --no-daemon -Dorg.gradle.welcome=never assemble` from the project root (you can override the command by setting `BP_GRADLE_BUILD_ARGUMENTS`), so you need to make sure that will result in building your JAR or WAR file. The buildpack will then look for your JAR or WAR at `build/libs/*.[jw]ar`. If that path is different for your project, then you need to set `BP_GRADLE_BUILT_ARTIFACT` and point it to the location of where it can expect your JAR or WAR. At this point, the same process as 1.) will happen, and the JAR or WAR is turned into an image.

The behavior above as well as all of the configuration options for working with Gradle are available [here](https://github.com/paketo-buildpacks/gradle/?tab=readme-ov-file#behavior).

**Reply by skagithala-clgx** [reply]
Hi Daniel,
[build.gradle files.zip](https://github.com/paketo-buildpacks/feedback/files/13986820/build.gradle.files.zip)

 
We are passing the entire project only and below is the command that we are using for doing the pack build
 
pack build <image-name>  --builder paketobuildpacks/builder-jammy-base  -e BP_GRADLE_BUILD_ARGUMENTS="clean build -Pversion=1.18.67 --console=plain --no-daemon --stacktrace --rerun-tasks --refresh-dependencies" -e BP_GRADLE_BUILD_FILE="./project1/" -e BP_GRADLE_ADDITIONAL_BUILD_ARGUMENTS="publish -x test  --stacktrace --info -Pversion=1.18.67 "
 
where project1 is the main project
 
Our project structure is as follows:
 
project1
project2
gradle
gradlew
gradlew.bat
settings.gradle
 
 
NOTE: project1 has dependency on project2. We are not using -p as the entire project is being passed into the container but it is not able to resolve the dependencies of project1. As project1 is dependent on project2 with reference to build.gradle at line number 96 of build.gradle for project1.
 
implementation project (':project2')
 
For reference we are attaching the build.gradle file for both project1 and project2
 
 
So could you please suggest if there is some changes required from config end.

**Reply by skagithala-clgx** [reply]
Hi Daniel,

Do we have any documentation or any similar issues faced on mono repo, please let us know.

**Reply by dmikusa** [reply]
Sorry, I don't know of any specific documentation for that task but there shouldn't be anything that's specific to buildpacks that would cause problems. 

The thing to keep in mind is that the buildpacks are just installing and running the tools for you. Think about this as if you would build your app in a container with a Dockerfile. What tools do you need in the container? What files do you need in the container? What commands need to be run in the container? 

The buildpacks will handle creating the isolated container for you and providing you with Java & Gradle. They will even run some default commands that work in many cases to make things easier. If the default commands don't work for your particular app then you need to change them to the commands that you would use to build your app. The doc link above has details about everything you can customize.

Hope that helps!
